### ISSUE TITLE
Investigate hoisting of bindings out lambda abstractions

### ISSUE THREAD

**Issue by jstolarek** [links-lang/links#593]
This tickets gathers some observations made during the two discussions between me, @slindley and @frank-emrich.  This is for future reference, since we're not planning to take any action at the moment.

One of the bugs fixed as part of work on the typed IR was incorrect handling of anonymous functions in the presence of quantified instantiation.  The problem was that during typechecking of an anonymous function it would be enclosed in a type abstraction (only when `quantified_instantiation` enabled, which was the default until #572):

```
TAbs Y. { fun (x) {x} }
```

`desugarFuns` would then convert this anonymous function to a named one:

```
TAbs Y. {
  fun [Z] freshfun (x : Z) {x}
  freshfun
}
```

The above code is incorrect because there is no application of type variable `Y` to `freshfun`.  So a correct version should be:

```
TAbs Y. {
  fun [Z] freshfun (x : Z) {x}
  TApp freshfun Y
}
```

This problem was originally reported as #446 and fixed with #534.

Later on, during compilation to IR, bindings such as the `freshfun` above are hoisted out of lambda abstractions.  @slindley made an observation that this might be unsound in general since we don't ensure that the type variables don't escape their scope.  In the example above, if `Y` appeared somewhere in `freshfun` it should be converted to an explicit type argument in the definition and then applied at every call site.  So basically lambda lifting, except that for type variables.

Now, the problem outlined above is hypothetical.  We have not seen this manifest as a bug anywhere so far and we don't know whether we generate programs that suffer from the outlined problem.  (At the moment type abstractions are only inserted by the compiler so there is no way for the user to write them explicitly - but this might change in the future.) A potential fix to this should be implemented in the `sugartoir` stage in the handling of type abstractions (`TAbstr`).  The major problem seems to be that currently hoisting and gathering of bindings is done by monadic plumbing and there is no immediate way to gain access only to the bindings hoisted out of the type abstraction at hand. Perhaps lifting of bindings could be augmented to accept a list of type variables to abstract over in the lifted bindings?

**Comment by jamescheney** [comment]
Yes, this sounds like what I initially thought we were going to have to do to resolve some of #446.

I think having either (a) a clear spec for what sugartoir does and how or (b) a reference implementation in a more direct form that we can use to generate well-typed IR would be helpful for fixing this and for maintainability.
